### DISCUSSION TITLE
EcoFlow dynamic charging adjustment

### DISCUSSION THREAD

**Discussion by apfelbuero** [svenerbe/ecoflow_dynamic_power_adjustment#1] [Ideas]
Since this script works so well for dynamic power adjustment of the PowerStream inverter, it would be a perfect opportunity to modify it for dynamic adjustment of the EcoFlow stations' charging level. This would allow us to adjust the AC charging power according to the excess PV generation. Maybe someone can help

**Comment by svenerbe** [comment]
Hi,
you mean the following: 
Lower limit settings for battery discharging(lowerLimit: 1-30) and Upper limit settings for battery charging(upperLimit: 70-100)? for the powerstream?

**Comment by campinoXL** [comment]
Hello,
Since I'm interested in this exact topic, I'll jump in.
I think it's about the charging power from a separate solar system. Depending on the solar surplus, the consumption of the EcoFlow should be regulated: if more power is produced than implemented, the surplus is used as completely as possible to charge the batteries.
The devices at the EcoFlow output should of course always be powered, if necessary directly via the input, but without charging the batteries - this should be possible via the 'backup reserve' settings (in the app under energy management).
Thanks for the great work so far!
Thomas

**Comment by apfelbuero** [comment]
Yes @campinoXL this is exactly what I meant. I did that already with a HA automation but it works not perfectly. The power stream script of @svenerbe works like a breeze!

**Reply by campinoXL** [reply]
Wie es aussieht (und wenn ich deinen Name richtig deute), sprechen wir hier momentan alle deutsch ...
Ich würde ehrlich gesagt nicht erwarten, dass Sven ein weiteres Skript 'nur' für unser Anliegen bastelt.
Vielleicht lässt du mir deine automation.yaml (oder zumindest die relevanten Ausschnitte) zukommen, dann würde ich auch mal in der Richtung probieren und müsste nicht bei Null anfangen.

**Reply by svenerbe** [reply]
hi @campinoXL, wie soll meine automation da helfen? Ich verstehe leider immer noch nicht euer anliegen bzgl. vielleicht könnt ihr mal erklären was ihr sucht, und welche attribute und welche ecoflow devices ihr steuern wollt. Meine automation hilft da denke nicht wirklich weiter. Grüße

**Reply by campinoXL** [reply]
hi @svenerbe, mich würde die Konfiguration von @apfelbuero interessieren, da er ja zumindest rudimentär etwas hinbekommen hat.
Was wir beide (ich könnte mir vorstellen: einige mehr) suchen, ist folgendes:
die überschüssige Energie, die von einem Balkonkraftwerk (das nicht an der EcoFlow angeschlossen ist) bereitgestellt wird, soll möglichst komplett zum Laden der Akkus genutzt werden.
Bei mir (EcoFlow River 2 Pro) klappt das im Moment nur von Hand über HA und die App:
ich schaue in HA, wie hoch das 'Minus' des Stromzählers ist, und regle entsprechend die Ladeleistung bzw. schalte das Laden überhaupt ein oder aus (über die App, Energiemanagement/Backup-Reserve, Schieber min/max).
Das könnte sicher mit einer Automatisierung laufen oder - und da kommst du ins Spiel ;-) - mit einem zusätzlichen Skript, das aus der Automatisierung mit Übergabe der 'freien' Leistung aufgerufen wird.
Gruß, Thomas

**Reply by apfelbuero** [reply]
Hallo zusammen. 

Ja genau das ist was wir suchen. Bei mir ist es so: Wir haben noch eine größere PV Anlage zusätzlich zum Balkonkraftwerk. Andere haben vielleicht schon ein BKW mit anderem Microwechselrichter und nutzen EcoFlows um ihre Kühlschränke oder andere Geräte zu betreiben. Nun wäre es perfekt die Ladeleistung in Watt der EcoFlows dem Überschuss des Balkonkraftwerks anzupassen. Es ist ja quasi genau das entgegengesetzte zu deinem PowerStream Skript. Die Anpassung sollte eigentlich recht simpel sein, ich hab nur leider gar keine Ahnung von Python.

Das Skript müsst also:
- den Überschuss am Energy Meter messen (statt dem Bedarf wie beim PowerStream Skript)
- die Ladeleistung auf der EcoFlow PowerStation entsprechend setzen (statt dem Leistungsausgang des PowerStream). Bei mir heisst die Enität z.B. "number.delta2_ac_charging_power"

Aktuell habe ich mir einen Helfer gebaut der "Netzbezug MINUS Netzeinspeisung" rechnet und diesen als "PowerOffset" definiert. 
Außerdem habe ich einen weiteren Helfer "Delta2_Soll" der berechnet sich aus "Delta2_AC_In MINUS PowerOffset" - also das was die PowerStation Delta 2 gerade aktuell an Leistung zieht, abzüglich dem PowerOffset. Das heisst wenn "PowerOffset" positiv ist geht die Ladeleistung runter - wenn PowerOffset negativ ist besteht Überschuss und die Ladeleistung geht hoch.

Leider klappt das nicht per Automation in HA - ich bekomme die Werte nicht in die EcoFlow geschrieben. Wenn ich die Werte manuell setze als Zahl, dann geht es aber. Daher die Idee, das per pyscript zu machen. 

Ich hänge heute Abend mal meine automation.yaml hier rein. Vielleicht bekommen wir es gemeinsam hin das in ein Script zu überführen.

Grüße, Valentin

**Reply by svenerbe** [reply]
Heisst du hasst dein anderes Balkonkraftwerk an der River 2 Pro am ac angeschlossen? Das wird denke ich so ohne weiteres nichts. Ich benutze ja die api von EcoFlow - die unterstützt die  River 2 Pro soweit nicht - siehe https://developer-eu.ecoflow.com/us/document/

**Reply by campinoXL** [reply]
Stimmt, die River 2 Pro hängt bei mir am Hausnetz.
Wenn die API sie nicht kennt, wäre das natürlich schade, von den topics in der API-Doku sieht die Delta 2 aber ähnlich aus (Anfrage an den EcoFlow-Support ist raus).
MIt HA-Bordmitteln (bzw. mit der EcoFlow Cloud Erweiterung) habe ich ähnliche Probleme wie @apfelbuero:
Wenn ich die Ladeleistung über die GUI einstelle, klappt alles, wenn ich das aber per Skript mache, bewegt sich der Schieber (z.B. von 100 W auf 200 W), der Wert kommt aber beim Gerät nicht an. Außerdem habe ich keinen Weg gefunden das Laden generell ein- und auszuschalten - das klappt wie gesagt nur über die App (Energiemanagement/Backup-Reserve).

**Reply by apfelbuero** [reply]
Also bei mir sind es Delta 2, die sich eigentlich über die API ganz prima ansteuern lassen.

**Comment by SweetHunter80** [comment]
Hi Sven,
bin bis zum Punkt Installation gekommen und habe das Script set_ef_powerstream_custom_load_power.py mit meine key und secret ersetzt.
Leider komme ich nicht mehr weiter.
Als nächster Punkt wäre nun Usage.
Wie muss ich weiter vorgehen?
Vielen Dank für deine Unterstützung.
Eddy

**Reply by svenerbe** [reply]
Hi Eddy, siehst du den neuen service wenn du einen neue automation erstellst - siehe usage!

**Reply by SweetHunter80** [reply]
Oder meinst du hier?
![image](https://github.com/svenerbe/ecoflow_dynamic_power_adjustment/assets/129670092/322010a3-013c-4ca5-8810-cd4acfac8d6a)

**Comment by SweetHunter80** [comment]
Hi Sven,
wenn ich auf Automation neu erstelle gehe, nach was soll ich suchen?
![image](https://github.com/svenerbe/ecoflow_dynamic_power_adjustment/assets/129670092/606cc6db-ba6f-4b0e-815b-f16176ecf157)

**Reply by svenerbe** [reply]
sorry, bitte erstmal ein paar Grundlagen ansehen. z.b. https://youtu.be/SGlkzKvJ96s?si=fTksB51jeIzH2RsY 
Auslöser, Bedingungen und Aktionen sollten schon bekannt sein.

ps: Über die Entwicklerwerkzeug kannst du den Dienst aber schon mal testen - da bei den Daten einfach das angegebene in Readme mit übergeben.
  
Grüße

**Reply by SweetHunter80** [reply]
Zieh ich mir mal heute Abend rein. Danke dir. ;-)

**Reply by SweetHunter80** [reply]
Grüß dich Sven,
habe nun bei der Automatisation im YAML Modus das reingemacht und auf Asuführen gecklickt.
Jedoch ändert er das Output nicht auf 10 Watt. Was mach ich falsch?
![image](https://github.com/svenerbe/ecoflow_dynamic_power_adjustment/assets/129670092/45353a19-6b7c-4bc9-8a7c-0a3ff51d70d7)

**Reply by SweetHunter80** [reply]
Es hat doch funktioniert nun hat er nach 30 sec. 20 Watt

**Reply by SweetHunter80** [reply]
hab 2 mal ausführen gedrückt. Anscheinend hat er die 2*10 Watt summiert.

**Reply by SweetHunter80** [reply]
scheint so als würde er doppeln bei jedem ausführen.

**Reply by svenerbe** [reply]
Moin, so soll es ja auch sein. Das Script addiert immer den Wert den du übergibst. Der Wert den du übergibst ist ja der von deinem total netzverbrauch aus deinem Haus. Z.b erster Lauf, netzverbrauch gerade 100 Watt, dann übergibst du 100 Watt, beim nächsten Lauf sollte dein netzvetbrauch 0 Watt  sein, dann übergibst du 0 Watt. Usw.. einfach auch mal ins Script nachlesen.

**Comment by SweetHunter80** [comment]
Hab’s kapert. Danke.Werde mal damit rumspielen.Mir ist auch aufgefallen das ich die automstisierung in der EcoFlow App aus haben muss sonnst wirkt er gegen.Als Tipp für andere bei der Fehlversuche.Vielen lieben Dank für deine Unterstützung.EddyGesendet mit der mobilen Mail AppAm 13.04.24 um 09:36 schrieb svenerbe
                
            
            
                
                    Von: "svenerbe" ***@***.***>Datum: 13. April 2024An: "svenerbe/ecoflow_dynamic_power_adjustment" ***@***.***>Cc: "SweetHunter80" ***@***.***>,"Comment" ***@***.***>Betreff: Re: [svenerbe/ecoflow_dynamic_power_adjustment] EcoFlow dynamic charging adjustment (Discussion #1)
                
                
Moin, so soll es ja auch sein. Das Script addiert immer den Wert den du übergibst. Der Wert den du übergibst ist ja der von deinem total netzverbrauch aus deinem Haus. Z.b erster Lauf, netzverbrauch gerade 100 Watt, dann übergibst du 100 Watt, beim nächsten Lauf sollte dein netzvetbrauch 0 Watt  sein, dann übergibst du 0 Watt. Usw.. einfach auch mal ins Script nachlesen.

**Reply by svenerbe** [reply]
Nice, kleiner tip. Ich bekomme bei mir alle 5 sec die Daten welcher Leistung aktuell vom Netz bezogen werden. Das benutze ich in meiner automation als trigger. Grüße

**Reply by SweetHunter80** [reply]
benutzt du ebenfalls eine Shelly3EM?

**Comment by SweetHunter80** [comment]
Bin gerade unterwegs. Kann ich heute Abend machen.Gesendet mit der mobilen Mail AppAm 13.04.24 um 13:48 schrieb campinoXL
                
            
            
                
                    Von: "campinoXL" ***@***.***>Datum: 13. April 2024An: "svenerbe/ecoflow_dynamic_power_adjustment" ***@***.***>Cc: "Comment" ***@***.***>,"SweetHunter80" ***@***.***>Betreff: Re: [svenerbe/ecoflow_dynamic_power_adjustment] EcoFlow dynamic charging adjustment (Discussion #1)
                
                
Wie es aussieht (und wenn ich deinen Name richtig deute), sprechen wir hier momentan alle deutsch ...
Ich würde ehrlich gesagt nicht erwarten, dass Sven ein weiteres Skript 'nur' für unser Anliegen bastelt.
Vielleicht lässt du mir deine automation.yaml (oder zumindest die relevanten Ausschnitte) zukommen, dann würde ich auch mal in der Richtung probieren und müsste nicht bei Null anfangen.

**Comment by SweetHunter80** [comment]
hier die Automatisation für die Nulleinspeisung:

alias: Nulleinspeisung
description: ""
trigger:
  - platform: numeric_state
    entity_id:
      - sensor.gesamtstrom
    for:
      hours: 0
      minutes: 0
      seconds: 5
    above: 5
condition: []
action:
  - service: pyscript.set_ef_powerstream_custom_load_power
    data:
      SerialNumber: die SN der Powerstream
      TotalPower: "{{ (states('sensor.powerstream_inverter_output_watts') | round(0)) }}"
      Automation: true
mode: single

**Reply by SweetHunter80** [reply]
Ist die erste Version, werde diese noch mit Batterielevel Regel usw. erweitern.

**Reply by SweetHunter80** [reply]
Irgendwie läuft das Script nicht?
Vielleicht wäre es möglich wenn Sven seine Automatisation veröffentlich könnte.

**Comment by SweetHunter80** [comment]
Hier eine funktionierende Auto:
alias: EcoFlow-PowerStream-dynamische-last
description: ""
trigger:
  - platform: time_pattern
    seconds: /10
condition:
  - type: is_battery_level
    condition: device
    device_id: XXXXXXXXXXXXXXXXXXXX
    entity_id: XXXXXXXXXXXXXXXXXXXXX
    domain: sensor
    above: 15
action:
  - service: pyscript.set_ef_powerstream_custom_load_power
    data:
      SerialNumber: HW51ZOHXXXXXXX
      TotalPower: "{{ (states('sensor.gesamtstrom') | float) | round }}"
      Automation: false
mode: single

**Reply by campinoXL** [reply]
cool, Danke!
eigentlich meinte ich aber apfelbuero - und das Alternativszenario "eigenständige Solaranlage - Überschussladen" (damit wurde der thread mal eröffnet) ;-) ...

**Comment by SweetHunter80** [comment]
Das Script funktioniert bis dato ganz gut. Danke an Sven.
Das einzige Manko was ich im Moment habe, das wenn meine Hausdach PV (diese ist nicht an die Powerstream/Powerstation angeschlossen) wieder genug Strom in mein Hausnetzt reinschiebt, Die Powerstream den wert der Einspeisung ins Netzt nicht auf 0 drückt.
Woran kann das liegen. am Script oder den Bedingungen?

**Reply by SweetHunter80** [reply]
Hier mal die Automatisation die ich im Moment am laufen habe:

alias: EcoFlow-PowerStream-dynamische-last
description: ""
trigger:
  - platform: time_pattern
    seconds: /5
condition:
  - condition: and
    conditions:
      - type: is_battery_level
        condition: device
        device_id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        entity_id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        domain: sensor
        above: 15
      - condition: numeric_state
        entity_id: sensor.gesamtstrom
        above: 10
action:
  - service: pyscript.set_ef_powerstream_custom_load_power
    data:
      SerialNumber: HW51ZOHXXXXXXXX
      TotalPower: "{{ (states('sensor.gesamtstrom') | float) | round }}"
      Automation: false
mode: single

**Reply by SweetHunter80** [reply]
Und das habe ich im Script angepasst:
       if not Automation:
            NewPermanentWatts= TotalPower *10

**Reply by svenerbe** [reply]
Moin, das kommt immer auf dein setup drauf an - ruf doch einfach in deiner automation wenn du keine dynamische leistungsanpassung mehr benötigst den service mit TotalPower: 0 auf, ich habe zum Beispiel ein helfer als boolean definiert, den setze ich wenn ich die dynamische anpassung brauche - wenn der nicht gesetzt ist, rufe ich den service wie folgt auf:
 
```
- service: pyscript.set_ef_powerstream_custom_load_power
        data:
          SerialNumber: SN
          TotalPower: 0
          Automation: false
```
Viele wege führen nach rom.... 
Weiß nicht ob wir hier jeden enzelnen step in einer automation beschreiben müssen. Grüße

**Reply by SweetHunter80** [reply]
Habs umgestellt und werde es beobachten. Thanks
